refactor(integ): restructure integration test framework

jusiskin opened the pull request:
## Changes

- run `PRE_AWS_INTERACTION_HOOK` between each stack deploy/destroy
- fix output redirection for parallel tests
- unify parallel and sequential code paths
- clean code organization
- clean log output and organize file artifacts. They now look like:

        .e2etemp/
        ├── amis.json
        ├── deadline_01_repository
        │   ├── deploy.txt
        │   ├── destroy.txt
        │   ├── exitcode
        │   ├── synth.log
        │   ├── test-output.txt
        │   ├── test-report.json
        │   └── timings.sh
        ├── deadline_02_renderQueue
        │   ├── deploy.txt
        │   ├── destroy.txt
        │   ├── exitcode
        │   ├── synth.log
        │   ├── test-output.txt
        │   ├── test-report.json
        │   └── timings.sh
        ├── deadline_03_workerFleetHttp
        │   ├── deploy.txt
        │   ├── destroy.txt
        │   ├── exitcode
        │   ├── synth.log
        │   ├── test-output.txt
        │   ├── test-report.json
        │   └── timings.sh
        └── deadline_04_workerFleetHttps
            ├── deploy.txt
            ├── destroy.txt
            ├── exitcode
            ├── synth.log
            ├── test-output.txt
            ├── test-report.json
            └── timings.sh
- update `integ/README.md` with updated instructions

## Testing

Ran the following combinations:

| Mode | Passing Tests | Failing Tests |
| ------ | ---------- | -------- |
| sequential | ✓ | ✓ |
| parallel | ✓ | ✓ |

For each ✓, this means:

*   All tests were executed
*   All test CloudFormation stacks are destroyed regardless of success/failure
*   In the case of a failed test
    *   the deployment and destroy logs are output
*   In the case of a successful test
    *   the deployment and destroy logs are not output
*   Test output is displayed at the bottom of the standard output stream

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

jusiskin committed:
refactor(integ): restructure integration test framework

- run `PRE_AWS_INTERACTION_HOOK` between each stack deploy/destroy
- fix output redirection for parallel tests
- unify parallel and sequential code paths
- clean code organization
- clean log output and organize file artifacts
- update README.md with updated instructions

ddneilson commented on integ/README.md:
```diff
@@ -3,22 +3,43 @@
 To run all test suites:
 
 1. Build and install dependencies by running build.sh from the top-level RFDK directory
+1. Configure AWS credentials (tests will use the default AWS profile, so either set up a default profile in `~/.aws/`
```
Instance role would be used as well, no? So there's a third option to temp creds or locally-stored creds

jusiskin replied:
Yes, I can elaborate on this to list and link to the suggested methods as documented in the [JS SDK Dev Guide](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html)

ddneilson commented on integ/README.md:
```diff
@@ -3,22 +3,43 @@
+1. Configure AWS credentials (tests will use the default AWS profile, so either set up a default profile in `~/.aws/`
+   credentials or use temporary credentials).
+1. Set the environment variable CDK_DEFAULT_REGION to the region the test should be deployed in
+1. Modify the `test-config.sh` configuration file. Currently the following options can be configured:
```
I'm surprised that we don't have an env var corresponding to: https://docs.aws.amazon.com/rfdk/api/latest/docs/aws-rfdk.deadline.ThinkboxDockerImages.html#userawsthinkboxeulaacceptance

jusiskin replied:
These are instructions for running the tests using the `test-config.sh` configuration file. All the variables in that file (including `USER_ACCEPTS_SSPL_FOR_RFDK_TESTS`) can be set using environment variables, but in order to avoid the config file from overriding the environment variables, an extra environment variable can be set:

```sh
export SKIP_TEST_CONFIG=true
```

EDIT: I'll update the README to document this tip

ddneilson replied:
Right. Sorry, I might not have been clear. We have an env var for the MongoDB SSPL, but I gather we don't have one for the Deadline license? Is the integ test just hard-coded to always accept that license?

ddneilson commented on integ/scripts/node/stack-order:
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env node
+require('./stack-order.js');
```
nit: missing newline

ddneilson commented on integ/scripts/bash/rfdk-integ-e2e.sh:
```diff
@@ -64,95 +69,51 @@ fi
-# Mark infrastructure destroy start time
-export INFRASTRUCTURE_DESTROY_START_TIME=$SECONDS
+# Run the component tests (potentially in parallel)
+get_component_dirs | xargs ${XARGS_ARGS} components/deadline/common/scripts/bash/component_e2e_driver.sh || cleanup_on_failure
```
Well damn, that's a nice simplification!

ddneilson reviewed (changes requested):
This is some great work Josh. I just have small nitpicky things.

jusiskin committed:
chore(integ): apply PR feedback

jusiskin committed:
fix(integ): invert backwards condition

horsmand commented on integ/scripts/bash/rfdk-integ-e2e.sh:
```diff
@@ -64,95 +69,51 @@ fi
+    -o                      \
+    -name "cdk.json"   |    \
+  # Filter out node_modules
+  grep -v node_modules |    \
```
I'm having trouble understanding why we care about `-path "*/node_modules"` and then filter them out. It seems like we're just ignoring the first half of the OR in the `find` command. Couldn't we just do `find . -name "cdk.json"`?

horsmand commented on integ/components/deadline/common/scripts/bash/component_e2e_driver.sh:
```diff
@@ -0,0 +1,45 @@
+# Fail on unset variables
+set -u
+
+# COMPONENT_NAME=$1
```
Possibly left in from testing.

horsmand commented on integ/scripts/node/stack-order.ts:
```diff
@@ -0,0 +1,278 @@
+/**
+ * The entrypoint of the program.
+ *
+ * This process and validates the cmomand line arguments. It exits and displays
```
```suggestion
 * This processes and validates the command line arguments. It exits and displays
```

horsmand commented on integ/scripts/bash/rfdk-integ-e2e.sh:
```diff
@@ -64,95 +69,51 @@ fi
+XARGS_ARGS="-n 1"
+if [[ "${RUN_TESTS_IN_PARALLEL-}" = true ]]
+then
+  # Unbounded parallelism
+  XARGS_ARGS="${XARGS_ARGS} -P 0"
```
Maybe expand on this comment to say that the `-P 0` flag will tell xargs to run all the commands in parallel and then block until they complete execution.

horsmand reviewed (commented):
My comments were really minor. This was quite the refactoring to wrap my head around but it will make both the code and the test results much easier to understand. The solution to refresh the credentials between stack deployments and test runs is simple but effective.

Also, I'm surprised how much work it was to determine the order of stacks in CDK, you'd think there would be a command built into their CLI for that.

jusiskin committed:
chore(integ): address CR feedback

- Fix typos and grammar
- Invert condition
- Fix hard-coded test result path
- Fix relative path issue
- Removed unused code
- Improve code comments
